### ISSUE TITLE
Investigate Calendar.ISO's dependency on `:calendar`

### ISSUE THREAD

**Issue by Qqwy** [elixir-lang/elixir#5887]
_This issue is so we won't forget about it. It only makes sense to work on this issue after #5841 has, in one way or another, been merged_.

### Environment

* Elixir & Erlang versions (elixir --version): 

Erlang/OTP 19 [erts-8.0] [source] [64-bit] [smp:8:8] [async-threads:10] [kernel-poll:false]

Elixir 1.5.0-dev (1cbc996cf)

### Current behavior

The current implementation of Calendar.ISO, and in extension Date, Time. DateTime and NaiveDateTime depend on the Erlang `:calendar` module for functions like `to_unix`, `from_unix`, `datetime_to_gregorian_days` and `datetime_from_gregorian_days`. For a reason unbeknownst to me (it is [not clearly documented](http://erlang.org/doc/man/calendar.html), `:calendar` does not work with dates before the [year 0](https://en.wikipedia.org/wiki/Year_0).

### Expected behavior

Work with ISO8601 dates (and datetimes) before the Year 0 without problems.

It is relatively easy to alter the implementations to use custom code written in Elixir instead. For `datetime_to_gregorian_days` and `datetime_from_gregorian_days`, these have already been written as part of the WIP-version of #5841. Conversion to/from UNIX is also very simple by using the new Rata Die intermediate format that #5841 will provide.

**Comment by josevalim** [comment]
I agree we can look in general to get rid of the `:calendar` dependency. Although the WIP proposal in #5841 is likely slower than the one in `:calendar`, so we need to look at some of the idioms in `:calendar` and use them on the new implementation, like removing the list traversals for number of days in years in favor of pattern matching and so on.

**Comment by Qqwy** [comment]
@josevalim Do you know if `:calendar` is implemented using native C-code? If it is not, I expect that most calculations are very similar to what was written in the elixir-code (which follows the code that @lau wrote, which follows the pseudocode outlined in Calendrical Calculations from Dershowitz and Rheingold).

**Comment by josevalim** [comment]
It is pure Erlang. It should be mostly similar except for pieces like this: https://github.com/erlang/otp/blob/master/lib/stdlib/src/calendar.erl#L534-L538

Which was implemented as a list in Elixir but it will be much faster as pattern matching.

**Comment by Qqwy** [comment]
I will start working on this :slightly_smiling_face: .

**Comment by ThomasArts** [comment]
While working on calendar and ISO8601.2004, the following observation might be relevant:
As a reference, I use: https://en.wikipedia.org/wiki/ISO_8601. There might be a newer version of 
ISO8601, but many people will be familiar with this version.

## Time module

###  The specification

Midnight is a special case and may be referred to as either "00:00" or "24:00". The notation "00:00" is used at the beginning of a calendar day and is the more frequently used. At the end of a day use "24:00". "2007-04-05T24:00" is the same instant as "2007-04-06T00:00".

### The implementation

```elixir
iex(1)> ~T[24:00:00]
** (ArgumentError) cannot parse "24:00:00" as time, reason: :invalid_time
    (elixir) lib/calendar.ex:498: Time.from_iso8601!/1
    (elixir) expanding macro: Kernel.sigil_T/2
             iex:3: (file)
```
There is a good reason to not allow 24:00:00, since it gives extra code for something that seems useless. Ok, the standard demands having this... but otherwise?
Well, there is this:

Write a generator for times between 15:00:00  and 23:59:59 feels unnatural. Moreover, it is not obvious that microseconds are taken into account, eg. is 23:59:59.2726 a time that you want in your test case?

Another point: theTime.compare function (which I would like to have as Time.before). Sooner or later, we want to add Date.before as well. It is often natural to say, the deadline is 2017-05-09 24:00:00, instead of 2017-05-10 00:00:00. But even just checking whether a time is between 18:00:00 and 24:00:00 is more natural than checking first it is between 18:00:00 and 23:59:59 and if not, whether it is 23:59:59 with some additional micro seconds.

**Proposal: if one does not use Erlang calendar module, then make 24:00:00 a special valid case in the new calendar module.** 
And do not store it as 00:00:00 internally! This would mess up comparison operators on date and time.

**Comment by josevalim** [comment]
Thank you @ThomasArts. Just to clarify, you said: "And do not store it as 00:00:00 internally!". Shouldn't it be stored as 00:00:00 on the next day internally? Otherwise we will need to add special handling almost everywhere. Or am I missing something?

**Comment by ThomasArts** [comment]
You can only save it internally as 00:00:00 if your Time data structure does also contain the date. Elixir version 1.4 that I am using has a time struct that does not contain the date. 

What date would you assign to ~T[08:35:04.1524] ? 0000-01-01 is probably fine, but kind of weird.
and the ~T[24:00:00] would match to 0000-01-02T00:00:00Z?

The Time structure should probably not contain the date, but then, 24:00:00 is a special case. I see the irritating amount of work and the amount of places where one can get it wrong! 

One alternative is to add a field :midnight to the struct, but I don't really like that, since there is an invariant that midnight can only be true if hour, minute, second and microseco
cond are zero. Hard to keep that invariant through all computations.

Question really is: does one want to be compatible with ISO8601? What happens if your Phoenix web server receives a JSON object with time-format as format field with content: "24:00:00" as time. (At the moment your validator will probably reject this valid ISO string). If we were to have a meeting over this from  2017-05-09T23:00:00+0900 to  2017-05-09T24:00:00+0900, then we might have a problem.

**Comment by josevalim** [comment]
> Question really is: does one want to be compatible with ISO8601? What happens if your Phoenix web server receives a JSON object with time-format as format field with content: "24:00:00" as time. (At the moment your validator will probably reject this valid ISO string).

Yes, on the other side, if Phoenix suddenly starts emitting dates with 24:00:00, it will probably break clients that do not correctly handle 24:00:00 (as we don't right now). Maybe we support 24:00:00 only in `Time` but always convert it to `00:00:00` if there is a `Date` attached?

@lau, what are your thoughts on this?

**Comment by ThomasArts** [comment]
I see a real problem in not supporting it for full. Then better (as today) not support it at all.

Assume you solve the following:
```elixir
iex(2)> meeting = ~T[23:30:00]
~T[23:30:00]
iex(3)> Time.compare(meeting, ~T[24:00:00]) 
:lt
```
Then this means that the programmar can use ~T"24:00:00" in any code.
It would be surprising if Term.to_iso8601(~T[24:00:00]) would result in "00:00:00". But I believe this could be right. It is then up to the programmer to handle the string in its context.

But the nice property Time.to_iso8601 (Time.from_iso8601(str)) == str 
would be violated for string "24:00:00". 
This is comparable with pretty printing of abstract forms. Given a string, parse it and pretty print it, one needs not get the same string back. That's fine. 

That's ok, as long as the more important property:
Time.from_iso8601 (Time.to_iso8601(time)) == time is still satisfied.
That is, given a parse tree, pretty print it and parse it, it should be the same parse tree.

However this property is going to be violated if 00:00:00 is printed and parsed again.
 
**Thus, if we do not print ~T[24:00:00] as 00:00:00 then we will in some code emit "24:00:00" to a system that may or may not be fully ISO8601 compatible.**

In other words, solving this is either agreeing with the ISO committee or taking place in that commitee and changing the standard.

**Comment by josevalim** [comment]
@ThomasArts, agreed. I believe my previous comment was confusing, let me rectify that:

For Time, we support 24:00:00, and we keep it as 24:00:00 everywhere. Including pretty printing, rendering and formatting.

For all others, 24:00:00 will be equivalent to 00:00:00 in the next day. This, however, may be an issue if timezones are involved and we should wait for @lau's feedback on that.

**Comment by ThomasArts** [comment]
@josevalim, sounds good to me. Make it a programmers responsibility to handle 24:00:00 with care when talking to other systems.

When a time enters a DateTime format, then one adjusts the date. This is pretty independent of time zones, I believe.

@lau Note that we love to have the following property
DateTime.from_iso8601 (DateTime.to_iso8601(time)) == time 
But that should be not problem there.... apart from the fact that the return value at the moment is not really a DateTime struct.

And we we are at it... can I suggest to rename DateTime.to_time to DateTime.time. The to_ prefix suggests a conversion and probably a _from operator, but there cannot be a _from.

**Comment by lau** [comment]
In Denmark there used to be a paper based ticketing system where the time could go past 24:00. 27:45 would mean 3:45 the next morning. It worked OK. The whole thing has been replaced with an overly complicated electronic ticket system "Rejsekort" which was a big scandal.

Even though 24:00 is in ISO8601 it is not really common to send JSON datetimes with 24:00 in them. How many issues has been posted because people wanted to parse 24:00 in any practical situation?

Using 24:00 or beyond is a special case that we can choose to not support in the standard library. People can still write their own parsers if they choose to support that.

<img src="https://upload.wikimedia.org/wikipedia/commons/8/8b/Klippekort.jpg" />

Imagine representing "2016-12-31T24:00:00" as being in 2016. It is the same time as 2017-01-01T00:00:00" so it is in 2017 and 2016 at the same time. Or is it? Do we want to represent it as 2016 even though it is really in 2017? The consequences sound like a nightmare. If we ask what year a date is in is it 2016 or 2017 or both? Should we return a list with valid years for one struct?

The simple solution is to not support 24:00 at all. And I think most users will not care at all and be happy that the standard library is not overly complicated.

**Comment by josevalim** [comment]
Thanks @lau. We will explicitly document this restriction in the docs then.

**Comment by ThomasArts** [comment]
Thank you so much @lau for this fantastic comment and picture. I love perspectives like these.

The standard is crippled, I fully agree. Documentating this is the right thing to do. Let's make the world a better place by removing legacy that is generating extra code, extra work and more unclarity.

**Comment by josevalim** [comment]
@Qqwy are you still planning to implement this? :)

**Comment by Qqwy** [comment]
Thanks for the heads-up!

Yes, I will implement this. I apologise for the delay; other things in my life required a lot of attention.

**Comment by Qqwy** [comment]
Do not worry, I am still planning to work on this in the very near future. Vacation is almost here... :sweat_smile: !

**Comment by Qqwy** [comment]
I am now working on it.

I am busy translating the relevant code of `:calendar` to Elixir, and comparing it to the Calendrical Calculations code that is written in Lisp (and thus picking the smarter one of the two when there is a difference between the two).

I've found out why `:calendar` does not allow dates below ISO year 0 (which is 'year 1 BC' in the proleptic Gregorian calendar as there was no definition of '0' when it was created).

The relevant code is here: https://github.com/erlang/otp/blob/master/lib/stdlib/src/calendar.erl#L439

It 'performs a recursive procedure' until the days match. The call to [`dy(year)`](https://github.com/erlang/otp/blob/master/lib/stdlib/src/calendar.erl#L517) (which is a cryptic name meaning `days_in_previous_years(year)`) does not work for years before `0`.

I think this is the major cause why `:calendar` does not support pre-0 dates. But there also is a secondary problem, which is that `:calendar` uses `div` and `rem` all over the place, which will not give the correct result for negative integers when used in this algorithm (we want floored division, not division-towards-zero).

**Comment by josevalim** [comment]
@Qqwy thanks! For now let's focus on keeping the same set of features and once the refactor is done we can discuss lifting those limitations!

**Comment by josevalim** [comment]
Closing in favor of the PR! :heart:
